module/VuFind/src/VuFind/Http/CachingDownloader.php

```diff
@@ -0,0 +1,168 @@
+    /**
+     * HTTP service
+     *
+     * @var HttpService
+     */
+    protected $httpService;
+
+    /**
+     * Constructor
+     *
+     * @param HttpService $httpService  HTTP service
+     * @param string      $cacheManager Base directory for cache
+     */
+    public function __construct(HttpService $httpService, CacheManager $cacheManager)
+    {
+        $this->httpService = $httpService;
+        $this->cacheManager = $cacheManager;
+        $this->setCacheId('downloader');
+    }
+
+    /**
+     * Set a different cache.
+     *
+     * @param string $cacheId Cache ID
+     *
+     * @return void
+     */
+    public function setCacheId($cacheId)
+    {
+        $cacheName = $this->cacheManager->addDownloaderCache($cacheId);
+        $this->cache = $this->cacheManager->getCache($cacheName);
+    }
+
+    /**
+     * Download a resource using the cache in the background.
+     *
+     * @param string   $url              URL
+     * @param array    $params           Request parameters (e.g. additional headers)
+     * @param callable $validateCallback Callback for validation
+     *                                   before storing to cache
+     * @param callable $decodeCallback   Callback for decoding
+     *
+     * @return mixed
+     */
+    public function download(
+        $url,
+        $params=[],
+        callable $validateCallback=null,
+        callable $decodeCallback=null
+    ) {
+        $cacheItemKey = md5($url);
+        foreach ($params as $paramKey => $paramValue) {
+            $cacheItemKey .= '#' . $paramKey . '§' . $paramValue;
+        }
+
+        // Add new item to cache if not exists
+        if (!$this->cache->hasItem($cacheItemKey)) {
+            $response = $this->httpService->get($url, $params);
+            if (!$response->isOk()) {
+                throw new \Exception('Could not download URL: ' . $url);
+            }
+            $body = $response->getBody();
+            if ($validateCallback !== null && ($validateCallback($body) === false)) {
+                throw new \Exception('Invalid response body from URL: ' . $url);
+            }
+            $this->cache->addItem($cacheItemKey, $body);
+        }
+
+        $body = $this->cache->getItem($cacheItemKey);
+        if ($decodeCallback === null) {
+            return $body;
+        } else {
+            return $decodeCallback($body);
+        }
+    }
+
+    /**
+     * Download a resource using the cache in the background,
+     * including validation / decoding for JSON.
+     *
+     * @param string $url    URL
+     * @param array  $params Request parameters (e.g. additional headers)
+     *
+     * @return stdClass
+     */
+    public function downloadJson($url, $params=[])
+    {
+        $validateJson = function ($json) {
+            // Use PhpSerialize instead of json_decode for better performance
+            $serializer = \Laminas\Serializer\Serializer::factory(
+                \Laminas\Serializer\Adapter\PhpSerialize::class
```

Comment:
Does this really work? Wouldn't you want to use the Json adapter to decode JSON? I don't understand how applying PHP unserialization to JSON data could work correctly.